src/lib/rules/http-compression/http-compression.ts

```diff
@@ -0,0 +1,672 @@
+
+import { Category } from '../../enums/category';
+import { getFileExtension, isTextMediaType } from '../../utils/content-type';
+import { getHeaderValueNormalized, isRegularProtocol, isHTTP, normalizeString } from '../../utils/misc';
+import { IAsyncHTMLElement, IResponse, IRule, IRuleBuilder, IFetchEnd } from '../../types';
+import { RuleContext } from '../../rule-context';
+import { CompressionCheckOptions } from './compression-check-options';
+
+const uaString = 'Mozilla/5.0 Gecko';
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+
+        const getRuleOptions = (property: string): CompressionCheckOptions => {
+            return Object.assign(
+                {},
+                {
+                    brotli: true,
+                    gzip: true,
+                    zopfli: true
+                },
+                (context.ruleOptions && context.ruleOptions[property])
+            );
+        };
+
+        const resourceOptions: CompressionCheckOptions = getRuleOptions('resource');
+        const targetOptions: CompressionCheckOptions = getRuleOptions('target');
+
+        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+        const checkIfBytesMatch = (rawResponse: Buffer, magicNumbers) => {
+            return rawResponse && magicNumbers.every((b, i) => {
+                return rawResponse[i] === b;
+            });
+        };
+
+        const getHeaderValues = (headers, headerName) => {
+            return (getHeaderValueNormalized(headers, headerName) || '').split(',');
+        };
+
+        const checkVaryHeader = async (resource, element, headers) => {
+            const varyHeaderValues = getHeaderValues(headers, 'vary');
+            const cacheControlValues = getHeaderValues(headers, 'cache-control');
+
+            if (!cacheControlValues.includes('private') &&
+                !varyHeaderValues.includes('accept-encoding')) {
+                await context.report(resource, element, `Should be served with the 'Vary' header containing 'Accept-Encoding' value.`);
+            }
+        };
+
+        const generateDisallowedCompressionMessage = (encoding: string) => {
+            return `Disallowed compression method: '${encoding}'.`;
+        };
+
+        const generateContentEncodingMessage = (encoding: string, notRequired?: boolean, suffix?: string) => {
+            return `Should${notRequired ? ' not' : ''} be served with the 'content-encoding${encoding ? `: ${encoding}` : ''}' header${suffix ? ` ${suffix}` : ''}.`;
+        };
+
+        const generateCompressionMessage = (encoding: string, notRequired?: boolean, suffix?: string) => {
+            return `Should${notRequired ? ' not' : ''} be served compressed${encoding ? ` with ${encoding}` : ''}${suffix ? ` ${suffix}` : ''}.`;
```

Comment:
We should make the message clearer.
Right now if a resources is not compressed the user is going to get 2 messages:

> Should be served compressed with Zopfli
> Should be served compressed with Brotli

which can be confusing. A way could be to say the protocol as you proposed the other day:

> Should be served compressed with Zopfli for HTTP
> Should be served compressed with Brotli for HTTPS

Reply:
Fixed.

(Note: Only changed the message for `Brotli`, since `Zopfli` applies to both)